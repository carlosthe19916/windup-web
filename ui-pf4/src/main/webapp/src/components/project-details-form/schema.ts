import { object, string } from "yup";

import { MigrationProject } from "models/api";
import { getProjectIdByName } from "api/api";

import { ProjectDetailsFormValues } from "./project-details-form";

export const projectDetailsFormInitialValue = (
  project?: MigrationProject
): ProjectDetailsFormValues => {
  return {
    name: project?.title || "",
    description: project?.description || "",
  };
};

export const projectDetailsFormSchema = (project?: MigrationProject) => {
  const validationSchema = object<ProjectDetailsFormValues>().shape({
    name: string()
      .trim()
      .required("This field is required.")
      .min(3, "The project name must contain at least 3 characters.")
      .max(120, "The project name must contain fewer than 120 characters.")
      .matches(
        /\s*[- \w]+\s*/,
        "The project name must contain only alphanumeric characters."
      )
      .test("uniqueValue", "The entered name is already in use.", (value) => {
        return getProjectIdByName(value!)
          .then(({ data }) => {
            const isValid: boolean =
              data === "" ||
              data === undefined ||
              data === null ||
              project?.title === value;
            return isValid;
          })
          .catch(() => false);
      }),
    description: string()
      .trim()
      .max(4096, "The description must contain fewer than 4096 characters."),
  });

  return validationSchema;
};
