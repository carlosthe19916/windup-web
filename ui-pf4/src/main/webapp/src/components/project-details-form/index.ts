export { ProjectDetailsForm } from "./project-details-form";
