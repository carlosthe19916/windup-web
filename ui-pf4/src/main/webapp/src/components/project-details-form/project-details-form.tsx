import * as React from "react";
import { Schema } from "@data-driven-forms/react-form-renderer";
import FormRenderer from "@data-driven-forms/react-form-renderer/dist/cjs/form-renderer";
import FormTemplate from "@data-driven-forms/pf4-component-mapper/dist/cjs/form-template";
import componentTypes from "@data-driven-forms/react-form-renderer/dist/cjs/component-types";
import validatorTypes from "@data-driven-forms/react-form-renderer/dist/cjs/validator-types";
import componentMapper from "@data-driven-forms/pf4-component-mapper/dist/cjs/component-mapper";

interface Form {
  name: string;
  description?: string;
}

export interface ProjectDetailsFormProps {
  initialValues?: Form;
  showFormControls?: boolean;
  searchProjectByName: (name: string) => Promise<any>;
  onSubmit: (value: any) => void;
  onCancel: () => void;
}

export const ProjectDetailsForm: React.FC<ProjectDetailsFormProps> = ({
  initialValues,
  showFormControls,
  searchProjectByName,
  onSubmit,
  onCancel,
}) => {
  const asyncNameValidator = (value: string) =>
    searchProjectByName(value)
      .then((response: any) => {
        if (response) {
          const error = { message: "The entered name is already in use" };
          throw error;
        }
      })
      .catch((error) => {
        throw error && error.message
          ? error.message
          : "There was an error validating 'name'";
      });

  const schema: Schema = {
    fields: [
      {
        name: "project",
        component: componentTypes.SUB_FORM,
        title: "Project details",
        fields: [
          {
            name: "name",
            label: "Name",
            helperText: "A unique name for the project",
            component: componentTypes.TEXT_FIELD,
            isRequired: true,
            validate: [
              asyncNameValidator,
              {
                type: validatorTypes.REQUIRED,
              },
              {
                type: validatorTypes.MIN_LENGTH,
                threshold: 3,
              },
            ],
          },
          {
            name: "description",
            label: "Description",
            helperText: "A short description of the project",
            component: componentTypes.TEXTAREA,
            isRequired: false,
            validate: [],
          },
        ],
      },
    ],
  };

  return (
    <FormRenderer
      schema={schema}
      componentMapper={componentMapper}
      FormTemplate={(props) => (
        <FormTemplate {...props} showFormControls={showFormControls} />
      )}
      onSubmit={onSubmit}
      onCancel={onCancel}
      initialValues={initialValues}
    />
  );
};
