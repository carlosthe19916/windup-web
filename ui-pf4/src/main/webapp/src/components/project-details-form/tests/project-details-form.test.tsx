import React from "react";
import { shallow } from "enzyme";
import { ProjectDetailsForm } from "../project-details-form";

describe("ProjectDetailsForm", () => {
  it("Renders without crashing", () => {});
});
