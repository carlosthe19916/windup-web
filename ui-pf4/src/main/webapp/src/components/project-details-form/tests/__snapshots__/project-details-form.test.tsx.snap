// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchTable Renders without crashing 1`] = `
<FormRenderer
  FormTemplate={[Function]}
  clearOnUnmount={false}
  componentMapper={
    Object {
      "checkbox": [Function],
      "date-picker": [Function],
      "dual-list-select": [Function],
      "field-array": [Function],
      "plain-text": [Function],
      "radio": [Function],
      "select": [Function],
      "slider": [Function],
      "sub-form": [Function],
      "switch": [Function],
      "tabs": [Function],
      "text-field": [Function],
      "textarea": [Function],
      "time-picker": [Function],
      "wizard": [Function],
    }
  }
  debug={[Function]}
  initialValues={Object {}}
  onCancel={[Function]}
  onSubmit={[Function]}
  schema={
    Object {
      "fields": Array [
        Object {
          "component": "sub-form",
          "fields": Array [
            Object {
              "component": "text-field",
              "helperText": "A unique name for the project",
              "isRequired": true,
              "label": "Name",
              "name": "name",
              "resolveProps": [Function],
              "validate": Array [
                [Function],
                Object {
                  "type": "required",
                },
                Object {
                  "threshold": 3,
                  "type": "min-length",
                },
              ],
            },
            Object {
              "component": "textarea",
              "helperText": "A short description of the project",
              "isRequired": false,
              "label": "Description",
              "name": "description",
              "validate": Array [],
            },
          ],
          "name": "project",
          "title": "Project details",
        },
      ],
    }
  }
/>
`;
